File: python-examples/02-structured-output/schemas.py
"""Pydantic schemas for structured output."""

from pydantic import BaseModel, Field


class MovieReview(BaseModel):
    """Structured output schema for movie reviews."""

    title: str = Field(description="The title of the movie")
    rating: int = Field(ge=1, le=10, description="Rating from 1-10")
    genre: str = Field(description="The movie's genre(s)")
    summary: str = Field(description="A brief summary of the movie")
    pros: list[str] = Field(description="List of positive aspects")
    cons: list[str] = Field(description="List of negative aspects")
    recommendation: str = Field(description="Who should watch this movie")


class RecipeOutput(BaseModel):
    """Structured output schema for recipes."""

    name: str = Field(description="Name of the recipe")
    prep_time_minutes: int = Field(description="Preparation time in minutes")
    cook_time_minutes: int = Field(description="Cooking time in minutes")
    servings: int = Field(description="Number of servings")
    difficulty: str = Field(description="Difficulty level: Easy, Medium, or Hard")
    ingredients: list[str] = Field(description="List of ingredients with quantities")
    instructions: list[str] = Field(description="Step-by-step cooking instructions")
    tips: list[str] = Field(default=[], description="Optional cooking tips")


class SentimentAnalysis(BaseModel):
    """Structured output schema for sentiment analysis."""

    text: str = Field(description="The analyzed text")
    sentiment: str = Field(description="Overall sentiment: positive, negative, or neutral")
    confidence: float = Field(ge=0.0, le=1.0, description="Confidence score 0-1")
    emotions: list[str] = Field(description="Detected emotions")
    key_phrases: list[str] = Field(description="Key phrases that influenced the analysis")

File: typescript-examples/02-structured-output/schemas.ts
/**
 * Zod schemas for structured output.
 */

import { z } from 'zod';

/** Structured output schema for movie reviews. */
export const movieReviewSchema = z.object({
  title: z.string().describe('The title of the movie'),
  rating: z.number().int().min(1).max(10).describe('Rating from 1-10'),
  genre: z.string().describe("The movie's genre(s)"),
  summary: z.string().describe('A brief summary of the movie'),
  pros: z.array(z.string()).describe('List of positive aspects'),
  cons: z.array(z.string()).describe('List of negative aspects'),
  recommendation: z.string().describe('Who should watch this movie'),
});

export type MovieReview = z.infer<typeof movieReviewSchema>;

/** Structured output schema for recipes. */
export const recipeOutputSchema = z.object({
  name: z.string().describe('Name of the recipe'),
  prep_time_minutes: z.number().int().describe('Preparation time in minutes'),
  cook_time_minutes: z.number().int().describe('Cooking time in minutes'),
  servings: z.number().int().describe('Number of servings'),
  difficulty: z.string().describe('Difficulty level: Easy, Medium, or Hard'),
  ingredients: z.array(z.string()).describe('List of ingredients with quantities'),
  instructions: z.array(z.string()).describe('Step-by-step cooking instructions'),
  tips: z.array(z.string()).default([]).describe('Optional cooking tips'),
});

export type RecipeOutput = z.infer<typeof recipeOutputSchema>;

/** Structured output schema for sentiment analysis. */
export const sentimentAnalysisSchema = z.object({
  text: z.string().describe('The analyzed text'),
  sentiment: z.string().describe('Overall sentiment: positive, negative, or neutral'),
  confidence: z.number().min(0).max(1).describe('Confidence score 0-1'),
  emotions: z.array(z.string()).describe('Detected emotions'),
  key_phrases: z.array(z.string()).describe('Key phrases that influenced the analysis'),
});

export type SentimentAnalysis = z.infer<typeof sentimentAnalysisSchema>;
